File: 11.Deteccion_de_Bordes/test_3.m
clc; clear; close all
pkg load image
A=imread('images/linea3.jpg');
%Convertir la imagen en binaria
B=im2double(A);
B(B<0.5)=0; B(B>=0.5)=1;
imshow(B)

%Calcular la discretización de theta y rho
%1. Theta: Tomar valores en [0,180] / [0, pi]
h1=0.1;
thetas=deg2rad(0:h1:180);
%2. Rho: Toma valores en [-d,d], donde d=sqrt(m^2+n^2), y [m,n]=size(B)
[m,n]=size(B);
d=sqrt(m^2+n^2);
h2=0.1;
rhos=-d:h2:d;

%Crear matriz de acumulación
Acumulador=zeros(length(thetas),length(rhos));

%Llenar la matriz de acumulación

[x_b,y_b]=find(B);

for i=1:length(x_b) %Recorrer los puntos del borde
  for theta_ind=1:length(thetas)
    theta=thetas(theta_ind);
    rho=x_b(i)*cos(theta)+y_b(i)*sin(theta);
    [~,rho_ind]=min(abs(rhos-rho));
    Acumulador(theta_ind,rho_ind)+=1;
  end  
end

%Mostrar graficamente el comportamiento del acumulador
%figure
%surface(thetas,rhos,Acumulador','EdgeColor','none')
%xlabel('rho')
%ylabel('theta')

lineas_intentos=10;

for r=1:lineas_intentos
  %Encontrar máxima posicion del Acumulador
  [xp,yp]=find(Acumulador==max(max(Acumulador)));
  %Observacion: Si el máximo se repite más de una vez, entonces 
  %             'x' y 'y' son vectores que tienen las posiciones

  %Graficar

  for k=1:length(xp)
    thetaMax=thetas(xp(k));
    rhoMax=rhos(yp(k));

    if abs(sin(thetaMax))<10^-4
      x_v=rhoMax/cos(thetaMax);
      line([n 1], [x_v x_v],'LineWidth',2)  
    else
      %Calcular pendiente
      pendiente=-cos(thetaMax)/sin(thetaMax);
      interseccion=rhoMax/sin(thetaMax);
      %Necesitamos el punto (1,y1)
      y1=pendiente*1+interseccion;
      %Necesitamos el punto (m,ym)
      ym=pendiente*m+interseccion;
      %Necesitamos el punto (x1,1)
      x1=(1-interseccion)/pendiente;
      %Necesitamos el punto (xn,n)
      xn=(n-interseccion)/pendiente;
      if pendiente>0
        if 0<y1
          line([y1 n], [1 xn],'LineWidth',2)  
        else
          line([1 ym], [x1 m],'LineWidth',2)  
        end    
      else
        if y1>m
          line([ym n], [m xn],'LineWidth',2)  
        else
          line([y1 1], [1 x1],'LineWidth',2)  
        end    
      end  
    end
    Acumulador(xp(k),yp(k))=0;
  end
end

File: 11.Deteccion_de_Bordes/test_3.py
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image

A = Image.open("images/linea3.jpg")
A = np.asarray(A).astype("float64")/255
A = A[:,:,0]
A[A>=0.5]=1
A[A<0.5]=0
plt.imshow(A, cmap=plt.cm.gray)

h1 = 1
thetas = np.deg2rad(np.arange(0,180+h1,h1))


m,n = np.shape(A)
d = np.sqrt(m**2 + n**2)
h2 = 1
rhos = np.arange(-d,d,h2)

Acumulador = np.zeros((len(thetas),len(rhos)))
x_borde,y_borde = np.nonzero(A)

for i in range(len(x_borde)):
    for theta_indice in range(len(thetas)):
        theta = thetas[theta_indice]
        rho = x_borde[i]*np.cos(theta) + y_borde[i]*np.sin(theta)
        rho_indice = np.argmin(np.abs(rhos-rho))
        Acumulador[theta_indice,rho_indice] += 1

lineas_intentos = 10

for r in range(lineas_intentos):
    #HAY UN PROBLEMA AQUI
    #XP RETORNA UN UNICO NUMERO Y DEBERIA RETORNAR UN ARRAY
    xp,yp = np.where(Acumulador == np.max(np.max(Acumulador)))

    y1 = 0
    xn = 0

    for k in range(len(xp)):
        thetaMax = thetas[k]
        rhoMax = rhos[k]

        if (np.abs(np.sin(thetaMax)) < 10**-5):
            xv = rhoMax / np.cos(thetaMax)
            plt.plot([n, 1], [xv, xv])
        else:
            pendiente = -np.cos(thetaMax) / np.sin(thetaMax)
            interseccion = rhoMax / np.sin(thetaMax)

            y1 = pendiente*1 + interseccion
            ym = pendiente*m + interseccion

            x1 = (1-interseccion)/pendiente
            xn = (n-interseccion)/pendiente

            if (pendiente > 0):
                if (0 < y1):
                    plt.plot([y1, n], [1, xn])
                else:
                    plt.plot([1, ym], [x1, m])
            else:
                if (y1 > m):
                    plt.plot([ym, n], [m, xn])
                else:
                    plt.plot([y1, 1], [1 , x1])

    plt.plot([y1, n], [1, xn])

plt.xlim([0,n])
plt.ylim([m,0])
plt.show() 